### PR TITLE
Replace backend_driver by new example runner.

### PR DESCRIPTION
**Pull request by anntzer**
backend_driver has been really outdated for a long time.  Replace it by
a new script running all or a subset of the examples scripts.

Supersedes #8069 (@dstansby).

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: tools/run_examples.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        self.elapsed = elapsed
+        self.failed = failed
+
+    def __str__(self):
+        s = ""
+        if self.backend:
+            s += f"{self.backend}: "
+        s += f"{self.elapsed}ms"
+        if self.failed:
+            s += "(failed!)"
```
Comment (QuLogic) [inline tools/run_examples.py:39]:
```suggestion
            s += " (failed!)"
```

File: tools/run_examples.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    parser.add_argument(
+        "--rundir", type=Path,
+        help=("directory from where the tests are run; defaults to a "
+              "temporary directory"))
+    parser.add_argument(
+        "paths", nargs="*", type=Path,
+        help="examples to run; defaults to all examples (except *_sgskip.py)")
+    args = parser.parse_args()
+
+    root = (Path(__file__).parent / "../examples").resolve()
```
Comment (QuLogic) [inline tools/run_examples.py:61]: `.parent` and `..`?

File: tools/run_examples.py (hunk 1 above)
Comment (anntzer) [reply tools/run_examples.py:39]: fixed

File: tools/run_examples.py (hunk 2 above)
Comment (anntzer) [reply tools/run_examples.py:61]: the first parent is basically dirname(); I guess the other option would be `.parents[1] / "examples"` but I always find it hard to get the right index in the brackets...

Comment (QuLogic) [reply tools/run_examples.py:61]: Could do `.parent.parent`, maybe?

Comment (anntzer) [reply tools/run_examples.py]: sure, that's fine too

**Comment by efiring** [comment]
Shouldn't it be possible to run this with no arguments?
```
(mpl1) ~/work/programs/py/mpl/matplotlib/tools $ python run_examples.py
Traceback (most recent call last):
  File "run_examples.py", line 91, in <module>
    main()
  File "run_examples.py", line 66, in main
    width = max(len(str(relpath)) for relpath in relpaths)
ValueError: max() arg is an empty sequence
```

**Comment by anntzer** [comment]
oops, should be fixed now.

**Pull request merged**